libutils/stack.c

```diff
@@ -0,0 +1,212 @@
+
+   This file is part of CFEngine 3 - written and maintained by Northern.tech AS.
+
+   This program is free software; you can redistribute it and/or modify it
+   under the terms of the GNU General Public License as published by the
+   Free Software Foundation; version 3.
+
+   This program is distributed in the hope that it will be useful,
+   but WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+   GNU General Public License for more details.
+
+  You should have received a copy of the GNU General Public License
+  along with this program; if not, write to the Free Software
+  Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA
+
+  To the extent this program is licensed as part of the Enterprise
+  versions of CFEngine, the applicable Commercial Open Source License
+  (COSL) may apply to this file if you as a licensee so wish it. See
+  included file COSL.txt.
+*/
+
+#include <alloc.h>
+#include <logging.h>
+#include <mutex.h>
+#include <stack.h>
+
+
+#define STACK_EMPTY       0
+#define EXPAND_FACTOR     2
+#define DEFAULT_CAPACITY 16
+
+/** @struct Stack_
+  @brief A simple thread-safe stack data structure.
+
+  Can push, pop, and copy. Also has functions for showing current stack size
+  and capacity, and if a stack is empty. If the amount of pushed elements
+  exceed the capacity, it will be multiplied by EXPAND_FACTOR and reallocated
+  with the new capacity. When destroying the stack, destroys each element with
+  the ItemDestroy function specified -- unless it is NULL -- and then proceeds
+  to destroy the lock, before freeing the data array and the stack itself.
+  */
+struct Stack_ {
+    pthread_mutex_t lock;             /**< Thread lock for accessing data. */
+    void (*ItemDestroy) (void *item); /**< Data-specific destroy function. */
+    void **data;                      /**< Internal array of elements.     */
+    int size;                         /**< Amount of elements in stack.    */
+    int capacity;                     /**< Current memory allocated.       */
+};
+
+static void DestroyRange(Stack *stack, int start, int end);
+static void ExpandIfNecessary(Stack *stack);
+
+Stack *StackNew(size_t initial_capacity, void (ItemDestroy) (void *item))
+{
+    Stack *stack = xmalloc(sizeof(Stack));
+
+    if (initial_capacity <= 0)
+    {
+        initial_capacity = DEFAULT_CAPACITY;
+    }
+
+    pthread_mutex_t lock = PTHREAD_ERRORCHECK_MUTEX_INITIALIZER_NP;
+    stack->lock = lock;
+
+    stack->capacity = initial_capacity;
+    stack->size = STACK_EMPTY;
+    stack->data = xcalloc(sizeof(void *), initial_capacity);
+    stack->ItemDestroy = ItemDestroy;
+
+    return stack;
+}
+
+void StackDestroy(Stack *stack)
+{
+    if (stack != NULL)
+    {
+        ThreadLock(&stack->lock);
+        DestroyRange(stack, STACK_EMPTY, stack->size);
+        ThreadUnlock(&stack->lock);
+    }
+
+    StackSoftDestroy(stack);
```

Comment:
I think there's a race condition here where something can push to the stack while it's being destroyed.

Reply:
I think this could be solved by just creating a new lock that is equal to the old one, and have the last thing you do being destroying the lock.

Reply:
There would still be a race condition, AFAIK, because you cannot destroy the lock while holding it.